src/serialization/ParseNodeProxyFactory.cs

```diff
@@ -54,5 +57,34 @@ public IParseNode GetRootParseNode(string contentType, Stream content)
             };
             return node;
         }
+        /// <summary>
+        /// Create a parse node from the given stream and content type.
+        /// </summary>
+        /// <param name="content">The stream to read the parse node from.</param>
+        /// <param name="contentType">The content type of the parse node.</param>
+        /// <param name="cancellationToken">The cancellation token for the task</param>
+        /// <returns>A parse node.</returns>
+        public async Task<IParseNode> GetRootParseNodeAsync(string contentType, Stream content, 
+            CancellationToken cancellationToken = default)
+        {
+            if (_concrete is not IAsyncParseNodeFactory asyncConcrete)
+            {
+                throw new Exception("IAsyncParseNodeFactory is required for async operations");
+            }
+            var node = await asyncConcrete.GetRootParseNodeAsync(contentType, content).ConfigureAwait(false);
+            var originalBefore = node.OnBeforeAssignFieldValues;
+            var originalAfter = node.OnAfterAssignFieldValues;
+            node.OnBeforeAssignFieldValues = (x) =>
+            {
+                _onBefore?.Invoke(x);
+                originalBefore?.Invoke(x);
+            };
+            node.OnAfterAssignFieldValues = (x) =>
+            {
+                _onAfter?.Invoke(x);
+                originalAfter?.Invoke(x);
+            };
```

Comment:
these assignments should be refactored to a method so they can be used in both overloads